crates/ark/src/interface.rs

```diff
@@ -1156,6 +1167,77 @@ impl RMain {
         }
     }
 
+    fn handle_kernel_request(&mut self, req: KernelRequest, info: &PromptInfo) {
+        log::trace!("Received kernel request {req:?}");
+
+        match req {
+            KernelRequest::EstablishUiCommChannel(ref ui_comm_tx) => {
+                self.handle_establish_ui_comm_channel(ui_comm_tx.clone(), info)
+            },
+        };
+    }
+
+    fn handle_establish_ui_comm_channel(
+        &mut self,
+        ui_comm_tx: Sender<UiCommMessage>,
+        info: &PromptInfo,
+    ) {
+        if self.ui_comm_tx.is_some() {
+            log::info!("Replacing an existing UI comm channel.");
+        }
+
+        // Create and store the sender channel
+        self.ui_comm_tx = Some(UiCommSender::new(ui_comm_tx));
+
+        // Go ahead and do an initial refresh
+        self.with_mut_ui_comm_tx(|ui_comm_tx| {
+            let input_prompt = info.input_prompt.clone();
+            let continuation_prompt = info.continuation_prompt.clone();
+
+            ui_comm_tx.send_refresh(input_prompt, continuation_prompt);
+        });
+    }
+
+    pub fn get_ui_comm_tx(&self) -> Option<&UiCommSender> {
+        self.ui_comm_tx.as_ref()
+    }
+
+    fn get_mut_ui_comm_tx(&mut self) -> Option<&mut UiCommSender> {
+        self.ui_comm_tx.as_mut()
+    }
+
+    fn with_ui_comm_tx<F>(&self, f: F)
+    where
+        F: FnOnce(&UiCommSender),
+    {
+        match self.get_ui_comm_tx() {
+            Some(ui_comm_tx) => f(ui_comm_tx),
+            None => {
+                // Trace level logging, its typically not a bug if the frontend
+                // isn't connected. Happens in all Jupyter use cases.
+                log::trace!("UI comm isn't connected.");
+            },
+        }
+    }
```

Comment:
The idea is that in `interface.rs` we use `with_ui_comm_tx()` quite a bit because it doesn't log an _error_ if the UI comm isn't connected - which is correct, because this code runs when in Jupyter mode too.

But in specific UI methods in other files (which only run in Positron mode), we use `get_ui_comm_tx()` which allows us to promote a `None` sender into an error, because that would be very unexpected